File: Sample/Global/UIManager.py
# -*- encoding: utf-8 -*-
r"""\brief UIManager: manages active UI canvases, event dispatch, and rendering order."""

from typing import Callable, List, Optional
from Engine.Utils import Math
from Engine.UI.Base import ControlBase, FunctionalBase
from .FocusManager import FocusGroup, FocusManager
from . import Manager
from .System import System


class UIManager:
    r"""\brief Manages active UI canvases, event dispatch, and rendering order.

    Handles loading, removal, and sorted update/render of UI canvases.
    """

    def __init__(self) -> None:
        r"""\brief Construct a UI manager."""
        self._UIs: List[ControlBase] = []
        self._focusManager: FocusManager = FocusManager()
        FunctionalBase.setKeyboardFocusResolver(self._focusManager.shouldDispatchKeyboardTo)
        FunctionalBase.setDirectionalFocusRequester(self._focusManager.requestDirectionalMove)
        FunctionalBase.setKeyboardFocusSetter(self._focusManager.setFocus)
        FunctionalBase.setKeyboardCursorResolver(self._focusManager.isCursorFocusOwner)

    def getFocusManager(self) -> FocusManager:
        r"""\brief Get the keyboard focus manager.

        - \return Focus manager owned by this UI manager.
        """
        return self._focusManager

    def setFocusNavigationEnabled(self, enabled: bool) -> None:
        r"""\brief Enable or disable focus-based keyboard navigation.

        - \param enabled True to route keyboard input by focused control.
        """
        self._focusManager.setNavigationEnabled(enabled)

    def registerFocusGroup(self, group: FocusGroup) -> None:
        r"""\brief Register an explicit focus group.

        - \param group Focus group to register.
        """
        self._focusManager.registerFocusGroup(group)

    @ExecSplit(default=(None,))
    def loadUI(self, ui: ControlBase) -> None:
        r"""\brief Load a UI canvas into the manager.

        - \param ui The canvas to load.
        """
        self._UIs.append(ui)
        if isinstance(ui, FunctionalBase):
            self._focusManager.registerElement(ui)

    @ReturnType(uis=List[ControlBase])
    def getUIs(self) -> List[ControlBase]:
        r"""\brief Get the list of all loaded UI canvases.

        - \return A list of ControlBase objects.
        """
        return self._UIs

    @ExecSplit(default=(None,))
    def removeUI(self, ui: ControlBase) -> None:
        r"""\brief Remove a UI canvas from the manager.

        - \param ui The canvas to remove.
        - \raises ValueError if the UI is not found.
        """
        if ui in self._UIs:
            self._UIs.remove(ui)
            if isinstance(ui, FunctionalBase):
                self._focusManager.unregisterElement(ui)
        else:
            raise ValueError("UI not found")

    def _fixedLogicHandle(self, fixedDelta: float) -> None:
        from Engine.UI import Canvas

        sortedUIs = sorted(
            self._UIs, key=lambda item: item.getZOrder() if isinstance(item, Canvas) else 0, reverse=True
        )
        for ui in sortedUIs:
            if ui.getVisible():
                if isinstance(ui, FunctionalBase):
                    ui.fixedUpdate(fixedDelta)

    def _logicHandle(self, deltaTime: float) -> None:
        from Engine.UI import Canvas

        FunctionalBase.setKeyboardFocusResolver(self._focusManager.shouldDispatchKeyboardTo)
        FunctionalBase.setDirectionalFocusRequester(self._focusManager.requestDirectionalMove)
        FunctionalBase.setKeyboardFocusSetter(self._focusManager.setFocus)
        FunctionalBase.setKeyboardCursorResolver(self._focusManager.isCursorFocusOwner)
        self._focusManager.prepareFrame()
        sortedUIs = sorted(
            self._UIs, key=lambda item: item.getZOrder() if isinstance(item, Canvas) else 0, reverse=True
        )
        for ui in sortedUIs:
            if ui.getVisible():
                if isinstance(ui, FunctionalBase):
                    ui.update(deltaTime)

    def _renderHandle(self, deltaTime: float, overlayRenderer: Optional[Callable[[], None]] = None) -> None:
        from Engine.UI import Canvas

        System.applyScreenTonePass()
        sortedUIs = sorted(self._UIs, key=lambda item: item.getZOrder() if isinstance(item, Canvas) else 0)
        for ui in sortedUIs:
            if ui.getVisible():
                if isinstance(ui, Canvas):
                    ui.render()
                System.draw(ui)
        if overlayRenderer is not None:
            overlayRenderer()
        System.display(deltaTime)
        for ui in sortedUIs:
            if ui.getVisible():
                if isinstance(ui, FunctionalBase):
                    ui.lateUpdate(deltaTime)

    def _updatePerformanceInfo(self, deltaTime: float) -> None:
        if not System.isDebugMode():
            return
        if Math.IsNearZero(Manager.TimeManager.getSpeed()):
            return

        realDeltaTime = deltaTime / Manager.TimeManager.getSpeed()
        FPS = 1.0 / realDeltaTime
        System.recordPerformance(FPS)

File: Sample/Core/Global/src/UIManager.cpp
#include <UIManager.hpp>

#include <System.hpp>
#include <Runtime/EngineState.hpp>
#include <UI/Canvas.hpp>
#include <UI/FunctionalBase.hpp>

#include <algorithm>
#include <stdexcept>

namespace {
int zOrder(const std::shared_ptr<ControlBase>& ui) {
    const std::shared_ptr<Canvas> canvas =
        std::dynamic_pointer_cast<Canvas>(ui);
    return canvas == nullptr ? 0 : canvas->getZOrder();
}

void renderCanvas(const std::shared_ptr<ControlBase>& ui) {
    const std::shared_ptr<Canvas> canvas =
        std::dynamic_pointer_cast<Canvas>(ui);
    sf::RenderTexture* target = System::getCanvas();
    if (canvas != nullptr && target != nullptr) {
        canvas->render(*target);
    }
}
}  // namespace

UIManager* UIManager::activeManager_ = nullptr;

UIManager::UIManager()
    : focusManager_(std::make_shared<FocusManager>()), displayScale_(Scale) {
    activateFocusResolvers();
}

UIManager::~UIManager() {
    if (activeManager_ != this) {
        return;
    }
    FunctionalBase::setKeyboardFocusResolver({});
    FunctionalBase::setDirectionalFocusRequester({});
    FunctionalBase::setKeyboardFocusSetter({});
    FunctionalBase::setKeyboardCursorResolver({});
    activeManager_ = nullptr;
}

void UIManager::shutdown() noexcept {
    FunctionalBase::setKeyboardFocusResolver({});
    FunctionalBase::setDirectionalFocusRequester({});
    FunctionalBase::setKeyboardFocusSetter({});
    FunctionalBase::setKeyboardCursorResolver({});
    activeManager_ = nullptr;
}

std::shared_ptr<FocusManager> UIManager::getFocusManager() const {
    return focusManager_;
}

void UIManager::setFocusNavigationEnabled(bool enabled) {
    focusManager_->setNavigationEnabled(enabled);
}

void UIManager::registerFocusGroup(const std::shared_ptr<FocusGroup>& group) {
    focusManager_->registerFocusGroup(group);
}

void UIManager::loadUI(const std::shared_ptr<ControlBase>& ui) {
    if (ui == nullptr) {
        throw std::invalid_argument("UI cannot be null");
    }
    ui->refreshDisplayScale();
    const std::lock_guard<std::mutex> lock(mutex_);
    uis_.push_back(ui);
    const std::shared_ptr<FunctionalBase> functional = functionalUI(ui);
    if (functional != nullptr) {
        focusManager_->registerElement(functional);
    }
}

std::vector<std::shared_ptr<ControlBase>> UIManager::getUIs() const {
    const std::lock_guard<std::mutex> lock(mutex_);
    return uis_;
}

void UIManager::removeUI(const std::shared_ptr<ControlBase>& ui) {
    const std::lock_guard<std::mutex> lock(mutex_);
    const auto iterator = std::find(uis_.begin(), uis_.end(), ui);
    if (iterator == uis_.end()) {
        throw std::invalid_argument("UI not found");
    }
    const std::shared_ptr<FunctionalBase> functional = functionalUI(*iterator);
    uis_.erase(iterator);
    if (functional != nullptr) {
        focusManager_->unregisterElement(functional);
    }
}

void UIManager::fixedLogicHandle(float fixedDelta) {
    const std::vector<std::shared_ptr<ControlBase>> sorted = sortedUIs(true);
    for (const std::shared_ptr<ControlBase>& ui : sorted) {
        if (!ui->getVisible()) {
            continue;
        }
        const std::shared_ptr<FunctionalBase> functional = functionalUI(ui);
        if (functional != nullptr) {
            functional->fixedUpdate(fixedDelta);
        }
    }
}

void UIManager::logicHandle(float deltaTime) {
    refreshDisplayScale();
    activateFocusResolvers();
    focusManager_->prepareFrame();
    const std::vector<std::shared_ptr<ControlBase>> sorted = sortedUIs(true);
    for (const std::shared_ptr<ControlBase>& ui : sorted) {
        if (!ui->getVisible()) {
            continue;
        }
        const std::shared_ptr<FunctionalBase> functional = functionalUI(ui);
        if (functional != nullptr) {
            functional->update(deltaTime);
        }
    }
}

void UIManager::refreshDisplayScale() {
    if (displayScale_ == Scale) {
        return;
    }
    displayScale_ = Scale;
    for (const std::shared_ptr<ControlBase>& ui : getUIs()) {
        if (ui != nullptr) {
            ui->refreshDisplayScale();
        }
    }
}

void UIManager::renderHandle(float deltaTime,
                             const std::function<void()>& overlayRenderer) {
    System::applyScreenTonePass();
    const std::vector<std::shared_ptr<ControlBase>> sorted = sortedUIs(false);
    for (const std::shared_ptr<ControlBase>& ui : sorted) {
        if (!ui->getVisible()) {
            continue;
        }
        renderCanvas(ui);
        System::draw(*ui);
    }
    if (overlayRenderer) {
        overlayRenderer();
    }
    System::composeFrame(deltaTime);
    for (const std::shared_ptr<ControlBase>& ui : sorted) {
        if (!ui->getVisible()) {
            continue;
        }
        const std::shared_ptr<FunctionalBase> functional = functionalUI(ui);
        if (functional != nullptr) {
            functional->lateUpdate(deltaTime);
        }
    }
}

std::vector<std::shared_ptr<ControlBase>> UIManager::sortedUIs(
    bool descending) const {
    std::vector<std::shared_ptr<ControlBase>> sorted;
    {
        const std::lock_guard<std::mutex> lock(mutex_);
        sorted = uis_;
    }
    std::stable_sort(sorted.begin(), sorted.end(),
                     [descending](const std::shared_ptr<ControlBase>& left,
                                  const std::shared_ptr<ControlBase>& right) {
                         return descending ? zOrder(left) > zOrder(right)
                                           : zOrder(left) < zOrder(right);
                     });
    return sorted;
}

void UIManager::activateFocusResolvers() {
    if (activeManager_ == this) {
        return;
    }
    activeManager_ = this;
    FunctionalBase::setKeyboardFocusResolver(
        [this](const FunctionalBase& element) {
            return focusManager_->shouldDispatchKeyboardTo(element);
        });
    FunctionalBase::setDirectionalFocusRequester(
        [this](FunctionalBase& element, const std::string& direction) {
            return focusManager_->requestDirectionalMove(element, direction);
        });
    FunctionalBase::setKeyboardFocusSetter([this](FunctionalBase& element) {
        return focusManager_->setFocus(element);
    });
    FunctionalBase::setKeyboardCursorResolver(
        [this](const FunctionalBase& element) {
            return focusManager_->isCursorFocusOwner(element);
        });
}

std::shared_ptr<FunctionalBase> UIManager::functionalUI(
    const std::shared_ptr<ControlBase>& ui) {
    return std::dynamic_pointer_cast<FunctionalBase>(ui);
}
